staging/src/k8s.io/apiserver/pkg/server/filters/with_early_late_annotations.go

```diff
@@ -0,0 +1,133 @@
+}
+
+func withShutdownLateAnnotation(handler http.Handler, shutdownInitiated lifecycleEvent, delayDuration time.Duration, clock utilsclock.PassiveClock) http.Handler {
+	if shutdownInitiated == nil || clock == nil || delayDuration == time.Duration(0) {
+		klog.Warning("WithShutdownLateAnnotation filter is not being added to the handler chain")
+		return handler
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		select {
+		case <-shutdownInitiated.Signaled():
+		default:
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		requestor, exists := request.UserFrom(req.Context())
+		if !exists {
+			responsewriters.InternalError(w, req, errors.New("no user found for request"))
+			return
+		}
+
+		defer handler.ServeHTTP(w, req)
+
+		var loopback bool
+		if requestor.GetName() == user.APIServerUser {
+			loopback = true
+		}
+
+		initiatedAt, _ := shutdownInitiated.SignaledAt()
+		elapsedSince := clock.Since(initiatedAt)
+		var veryLate bool
+		// TODO: 80% is the threshold, if requests arrive after 80% of
+		// shutdown-delay-duration elapses we annotate the request as late=true.
+		if float64(elapsedSince)/float64(delayDuration) > 0.80 {
+			veryLate = true
+		}
+
+		value := fmt.Sprintf("since=%s elapsed=%s shutdown-delay-duration=%s loopback=%t late=%t",
+			shutdownInitiated.Name(), elapsedSince.String(), delayDuration, loopback, veryLate)
+		audit.AddAuditAnnotation(req.Context(), "apiserver.k8s.io/shutdown", value)
+		if !loopback && veryLate {
+			klog.Warningf("Late request (80%% past shutdown-delay-duration) "+
+				"URI: %q, RemoteAddr: %q, UserAgent: %q, User: %s, shutdown-delay-duration: %s, elapsed: %s, loopback: %t",
+				req.RequestURI, req.RemoteAddr, req.UserAgent(), requestor.GetName(), delayDuration, elapsedSince, loopback)
+		}
+	})
+}
+
+// WithStartupEarlyAnnotation, if added to the handler chain, tracks the
+// incoming request(s) that arrive at the apiserver before it is ready to
+// serve request(s), and annotates the audit event for these request(s) with
+// diagnostic information.
+// This enables us to identify the actor(s)/load balancer(s) that are sending
+// requests too early to the apiserver.
+func WithStartupEarlyAnnotation(handler http.Handler, hasBeenReady lifecycleEvent) http.Handler {
+	if hasBeenReady == nil {
+		klog.Warning("WithStartupEarlyAnnotation filter is not being added to the handler chain")
+		return handler
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		select {
+		case <-hasBeenReady.Signaled():
+			handler.ServeHTTP(w, req)
+			return
+		default:
+		}
+
+		requestor, exists := request.UserFrom(req.Context())
+		if !exists {
+			responsewriters.InternalError(w, req, errors.New("no user found for request"))
+			return
+		}
+
+		defer handler.ServeHTTP(w, req)
+
+		if requestor.GetName() == user.APIServerUser {
+			return
+		}
+		audit.AddAuditAnnotation(req.Context(), "apiserver.k8s.io/startup", "early=true loopback=false")
```

Comment:
how did you know loopback is false?

Reply:
i meant `loopback` as apiserver sending request to itself, i fixed it, now `self` means apiserver sending request to itself and `loopback` means loopback IP.